### DISCUSSION TITLE
Change chart's position

### DISCUSSION THREAD

**Discussion by Harshpanday** [mesa/mesa#1364] [Q&A]
Hey everyone, 
So I want to change the position of the chart that appears on the web app.

![image](https://user-images.githubusercontent.com/48926152/173259016-969139b1-b414-44ff-a3ae-c65273527c84.png)

As pointed out in the above image, I want to change the position of the chart from the bottom to the left side.
I am looking through the root code but cannot find where I should make changes to get the desired result.
Any suggestions are greatly appreciated.

**Comment by rht** [comment]
You can modify the chart CSS in https://github.com/projectmesa/mesa/blob/99a71c02d14dd1f2e3426b84c3207faf832f8d7a/mesa/visualization/templates/js/ChartModule.js#L6 with `display: inline-block` or something.
This is for the map: https://github.com/projectmesa/mesa-geo/blob/419f395ee953c8bbbd902d04de0a4aad840406b7/mesa_geo/visualization/templates/js/MapModule.js#L3

**Reply by Harshpanday** [reply]
Thank you for the quick reply,
I forked mesa and made changes, previously I was working with version 0.8.9.
Now I have updated mesa, but after updating my Barchart disappeared. 

![image](https://user-images.githubusercontent.com/48926152/173423843-272e95d4-90f8-48b4-a7de-0783244ccc82.png)

I am importing it ```from mesa.visualization.modules import BarChartModule``` , I am not getting any error, just the visualization completely disappeared other than that the ABM is working fine.
I'll be very grateful if you could let me know what am I doing wrong.

My complete server code is 
```
from mesa_geo.visualization.ModularVisualization import ModularServer
from mesa.visualization.modules import BarChartModule

from mesa.visualization.UserParam import UserSettableParameter
from main import ABM,erhc,erlc,lrhc,lrlc,cspm,spm
from mesa_geo.visualization.MapModule import MapModule

def abm_draw(agent):
    potrayal = dict()

    #if agent.atype is None:
        #potrayal["color"] = "Grey"
    #elif agent.atype == 0:
        #potrayal["color"] = "Red"
    if type(agent) is erhc:
        potrayal["color"] = "Red"
    if type(agent) is erlc:
        potrayal["color"] = "Blue"
    if type(agent) is lrhc:
        potrayal["color"] = "Green"
    if type(agent) is lrlc:
        potrayal["color"] = "Yellow"
    if type(agent) is spm:
        potrayal["color"] = "Black"
    elif type(agent) is cspm:
        potrayal["color"] = "Purple"
    return potrayal

#happy_element = HappyElement()
map_element = MapModule(abm_draw,ABM.Map_coordinates, 9,700,700)
model_params = { }

abm_chart = BarChartModule(
    [{"Label": "ERHC_FA", "Color": "Red"},{"Label": "ERLC_FA", "Color": "Blue"},
     {"Label": "LRHC_FA", "Color": "Green"},{"Label": "LRLC_FA", "Color": "Yellow"}],canvas_height=300, canvas_width=500,
    data_collector_name="datacollector"

)

server = ModularServer(
    ABM, [map_element,abm_chart], "ABM",model_params
)
```

**Reply by rht** [reply]
You need to check the browser console to see if there is any error message there.

**Reply by Harshpanday** [reply]
Ohh, I didn't see that.
It says ```d3 is not defined at new BarChartModule```.
do I need to define/assign anything to the graph?

**Reply by rht** [reply]
You probably need to fork from the 0.9.0 release tag instead of latest main of Mesa. There have been several backward incompatible changes in main. @wang-boyu just wondering if you see any issue with latest main of Mesa with Mesa-Geo.

**Reply by Harshpanday** [reply]
Thank you for the help.
One last thing, I am making changes to the root code and pushing them to my code and then I am installing mesa from my latest push.
Since I am using the Barchart I am making changes to ```BarChartModule.js```.
![image](https://user-images.githubusercontent.com/48926152/173464349-c3828732-540d-48f0-a942-b0aa31e1a3b0.png)

But nothing is changing in my visualization even after reinstalling mesa from the latest push.
For example the above style tag is empty instead of having ```border:1px dotted``` inside it but my visualization still shows dotted border for the graph
![image](https://user-images.githubusercontent.com/48926152/173464550-fb23e809-ca31-4420-868c-1a04b53b02b5.png)

Am I making the changes at the wrong place or am I not editing the root code correctly?

**Reply by rht** [reply]
Try `pip install -e .` instead of the usual `pip install .`. That makes it so that any change you make locally is automatically propagated to the installed library without having to reinstall.

**Reply by Harshpanday** [reply]
I did ```pip install -e```.
still, after making changes nothing is showing in the visualization.

**Reply by rht** [reply]
Try a more drastic change, e.g. removing the svg tag completely.

**Reply by Harshpanday** [reply]
No difference in visualization even after removing the complete ```svg tag```

**Reply by wang-boyu** [reply]
> You probably need to fork from the 0.9.0 release tag instead of latest main of Mesa. There have been several backward incompatible changes in main. @wang-boyu just wondering if you see any issue with latest main of Mesa with Mesa-Geo.

I'm guessing it's due to `modular_template.html`. It is copied from `Mesa` to `Mesa-Geo` during setup, but not yet updated with latest changes in `Mesa`.

The question is, should `Mesa-Geo`'s master branch be updated with latest `Mesa` from pypi or source code?

**Reply by rht** [reply]
I see, that explains it. IMO it should always be from an explicit version in PyPI, not Git latest main.

Does it copy from any Mesa version that the user already installs in their machine? Does this mean @Harshpanday just need to reinstall Mesa-Geo whenever he changes his Mesa installation?

**Reply by rht** [reply]
I think we need a more orthogonal solution where a change in the user's local Mesa lib is immediately picked up by Mesa-Geo. This means we probably can't use the copying mechanism.

**Reply by wang-boyu** [reply]
The current setup copies from lastest pypi Mesa. If @Harshpanday installed with `pip install -e .`, then there should be a bunch of js/css files that are copied over and ignored by git, in `mesa_geo/visualization/templates` folder. Subsequent changes needs to be done to the files in this folder (if I'm not wrong).

Given the current active changes to Mesa, I'm thinking to have some way to keep Mesa-Geo updated accordingly. Perhaps we can have a rc (release candidate) branch in Mesa-Geo that is updated with Mesa source code. In this way whenever there's a release in Mesa, we can merge rc to master and make a release.

**Reply by wang-boyu** [reply]
> I'm guessing it's due to `modular_template.html`. It is copied from `Mesa` to `Mesa-Geo` during setup, but not yet updated with latest changes in `Mesa`.

My previous comment is wrong - during setup, everything else is copied over from Mesa 0.9.0 other than `modular_template.html`.

**Reply by rht** [reply]
A better solution is to remove the copying mechanism and use the files in the install Mesa library folder.

**Reply by rht** [reply]
^clarification: remove the copying mechanism, and instead, dynamically use the files in the installed Mesa library folder.

**Reply by wang-boyu** [reply]
The main challenge is that Mesa-Geo also uses a copy of `ModularServer` from `ModularVisualization.py`, so that `leaflet.js` and `leaflet.css` can be loaded as `package_includes`, and uses can define their own `local_includes`.

As a result, Mesa-Geo should only import `ModularServer` from `mesa_geo.visualization.ModularVisualization`, not `mesa`, which could be confusing.

We can probably avoid this by removing the copying mechanism, but I'm not sure how to make `package_includes` and `local_includes` work in this case.

**Reply by rht** [reply]
Still doable. You can definitely inherit from Mesa's `ModularServer`, and then patch into its `package_includes` the absolute path of leaflet.js and leaflet.css.

**Reply by Corvince** [reply]
Yes I would actually solve this on the side of mesa by improving the way Js and css modules are loaded. I don't think it makes sense to have an independent modular server. Or maybe it would make sense to just expose the absolute paths if the leaflet files and include them as local includes

**Reply by rht** [reply]
Another solution without involving path is to inject the string content of leaflet.js and leaflet.css into `ModularServer`'s include. This solves a potential future issue with Jupyter not allowing you to load from any arbitrary filesystem paths.

**Reply by wang-boyu** [reply]
Would inheriting Mesa's `ModularServer` still require users to import the derived class from Mesa-Geo?

The main issue to me is that Mesa is open for users to have their own `local_includes`, but not so flexible for extensions/plugins. Imagine another person develops another Mesa extension with his/her own `ModularServer`. In this case how to use Mesa-Geo and this new extension together?

It would be better to only import Mesa-Geo for GIS related classes, and use Mesa's own core components like `ModularSever`. For instance in NetLogo we can do `extensions [gis]`, then the gis functions will be available in the `gis:` namespace. Similarly we can load other extensions such as `extensions [nw]`.

But I guess this plugin mechanism would need lots of effort to implement : )

**Reply by rht** [reply]
We can use vanilla Mesa's `ModularServer` if we can do this
```python
import mesa
import mesa_geo
server = mesa.visualization.ModularServer(
    ConwaysGameOfLife, [canvas_element], "Game of Life", {"height": 50, "width": 50},
    extensions=[mesa_geo.modular_server_extensions],
)
```
`ModularServer` will then either deduce the full path to the leaflet files from `mesa_geo.modular_server_extensions`, or the entire string content of the leaflet files as inline JS/CSS.

**Reply by rht** [reply]
For the lazy, you can define `mesa_geo.visualization.ModularServer` that prepopulates the `extensions` argument.

**Reply by wang-boyu** [reply]
This is great. Could you define an interface for `mesa_geo.modular_server_extensions` in Mesa, so that I can implement it in Mesa-Geo? Like you said, could be file path or string content.

Then we can get rid of `mesa_geo.visualization.ModularServer` or any `ModularServer` from future plugins of Mesa.

**Reply by rht** [reply]
I have decided to use the absolute path because having inline JS/CSS when you do "view source" in the browser is nasty.
